src/plot.cpp

```diff
@@ -1569,6 +1524,303 @@ void ProjectionPlot::set_field_of_view(pugi::xml_node node)
+
+    if (light_pos_tmp.size() != 3)
+      fatal_error("Light position must be given as 3D coordinates");
+
+    light_location_.x = light_pos_tmp[0];
+    light_location_.y = light_pos_tmp[1];
+    light_location_.z = light_pos_tmp[2];
+  } else {
+    light_location_ = camera_position();
+  }
+}
+
+void PhongPlot::set_diffuse_fraction(pugi::xml_node node)
+{
+  if (check_for_node(node, "diffuse_fraction")) {
+    diffuse_fraction_ = std::stod(get_node_value(node, "diffuse_fraction"));
+    if (diffuse_fraction_ < 0.0 || diffuse_fraction_ > 1.0) {
+      fatal_error("Must have 0<=diffuse fraction<= 1");
+    }
+  }
+}
+
+void Ray::compute_distance()
+{
+  dist_ = distance_to_boundary(*this);
+}
+
+void Ray::trace()
+{
+  int first_surface_ = -1;    // surface first passed when entering the model
+  first_inside_model_ = true; // false after entering the model
+  while (intersection_found_) {
+    bool inside_cell = false;
+
+    i_surface_ = std::abs(surface()) - 1;
+
+    // This means no surface was intersected. See cell.cpp
+    // and search for numeric_limits to see where we return it.
+    if (surface() == std::numeric_limits<int>::max()) {
+      warning(fmt::format("Lost a ray, r = {}, u = {}", r(), u()));
+      return;
+    }
+
+    if (i_surface_ > 0 &&
+        model::surfaces[i_surface_]->geom_type_ == GeometryType::DAG) {
+#ifdef DAGMC
+      int32_t i_cell = next_cell(
+        i_surface_, cell_last(n_coord() - 1), lowest_coord().universe);
+      inside_cell = i_cell >= 0;
+#else
+      fatal_error("Not compiled for DAGMC, but somehow you have a DAGCell!");
+#endif
+    } else {
+      inside_cell = exhaustive_find_cell(*this, settings::verbosity >= 10);
+    }
+
+    if (inside_cell) {
+
+      hit_something_ = true;
+      intersection_found_ = true;
+
+      compute_distance();
+
+      if (first_inside_model_) {
+        i_surface_ = first_surface_ - 1;
+        first_inside_model_ = false;
+      }
+
+      // Call the specialized logic for this type of ray
+      on_intersection();
+      if (stop_)
+        return;
+
+      // There are no more intersections to process
+      // if we hit the edge of the model, so stop
+      // the particle in that case:
+      if (dist_.distance == INFTY || dist_.distance == INFINITY) {
+        return;
+      }
+
+      // Advance particle, prepare for next intersection
+      for (int lev = 0; lev < n_coord(); ++lev) {
+        coord(lev).r += dist_.distance * coord(lev).u;
+      }
+      surface() = dist_.surface_index;
+      n_coord_last() = n_coord();
+      n_coord() = dist_.coord_level;
+      if (dist_.lattice_translation[0] != 0 ||
+          dist_.lattice_translation[1] != 0 ||
+          dist_.lattice_translation[2] != 0) {
+        cross_lattice(*this, dist_, settings::verbosity >= 10);
+      }
+
+    } else {
+      first_surface_ = RayTracePlot::advance_to_boundary_from_void(*this);
+      intersection_found_ = first_surface_ != -1; // -1 if no surface found
```

Comment:
Can we simply return here?

Reply:
No, because this advances forward to the next surface boundary it hits. It gets called repeatedly until we're in the geometry.

Reply:
Ah, understood. Thanks!